### PR TITLE
feat: add support for multiple tunnel destinations in tailnet

### LINKED ISSUE #14729: enhance tailnetAPIConnector to support multiple peers
Right now, the `tailnetAPIConnector` used for client connections only supports a single peer (a workspace Agent), specified at instantiation.  For the use case of a native desktop VPN, we need to support multiple peers that can be added and removed as workspaces are started and stopped.

In implementing this, we should consider refactoring to share code between `servertailnet` and `tailnetAPIConnector`, both of which need to support multiple Agent peers.

### PR DESCRIPTION
**Pull request by spikecurtis**
Closes #14729

Expands the Coordination controller used by the CLI client to allow multiple tunnel destinations (agents).  Our current client uses just one, but this unifies the logic so that when we add Coder VPN, 1 is just a special case of "many."

### CODE REVIEW

**Comment by spikecurtis** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/15409?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15409** <a href="https://app.graphite.dev/github/pr/coder/coder/15409?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#15408** <a href="https://app.graphite.dev/github/pr/coder/coder/15408?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15361** <a href="https://app.graphite.dev/github/pr/coder/coder/15361?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15347** <a href="https://app.graphite.dev/github/pr/coder/coder/15347?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15346** <a href="https://app.graphite.dev/github/pr/coder/coder/15346?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15345** <a href="https://app.graphite.dev/github/pr/coder/coder/15345?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15344** <a href="https://app.graphite.dev/github/pr/coder/coder/15344?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15343** <a href="https://app.graphite.dev/github/pr/coder/coder/15343?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: tailnet/controllers.go [hunk 1]
```diff
@@ -287,34 +288,139 @@ func (c *BasicCoordination) respLoop() {
...
+		}
+		c.coordination = nil
+	}
+}
+
+func (c *TunnelSrcCoordController) SyncDestinations(destinations []uuid.UUID) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	toAdd := make(map[uuid.UUID]struct{})
+	toRemove := maps.Clone(c.dests)
```
Comment (ethanndickson) [inline tailnet/controllers.go:382]: Could use `slice.SymmetricDifference`?

**Review by ethanndickson — APPROVED** [review]
I like these tests, easy to follow.
LGTM 👏

File: tailnet/controllers.go [hunk 2]
```diff
@@ -287,34 +288,139 @@ func (c *BasicCoordination) respLoop() {
...
+				c.Logger.Debug(context.Background(),
+					"failed to close coordinator client during sync destinations",
+					slog.Error(cErr))
+			}
+			c.coordination = nil
+		}
+	}()
+	for dest := range toAdd {
+		err = c.coordination.Client.Send(
+			&proto.CoordinateRequest{AddTunnel: &proto.CoordinateRequest_Tunnel{Id: dest[:]}})
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply tailnet/controllers.go:410]: There is some risk, yeah.  In practice we never mutate a UUID, but it's probably better to be sure.  I'll fix.

File: tailnet/controllers.go (hunk 1 above)
Comment (spikecurtis) [reply tailnet/controllers.go:382]: `c.dests` is not a slice.

**Pull request merged**